packages/firestore/src/local/memory_document_overlay_cache.ts

```diff
@@ -0,0 +1,183 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { DocumentKey } from '../model/document_key';
+import { Mutation } from '../model/mutation';
+import { Overlay } from '../model/overlay';
+import { ResourcePath } from '../model/path';
+import { SortedMap } from '../util/sorted_map';
+
+import { DocumentOverlayCache } from './document_overlay_cache';
+import { PersistencePromise } from './persistence_promise';
+import { PersistenceTransaction } from './persistence_transaction';
+
+/**
+ * An in-memory implementation of DocumentOverlayCache.
+ */
+export class MemoryDocumentOverlayCache implements DocumentOverlayCache {
+  // A map sorted by DocumentKey, whose value is a pair of the largest batch id
+  // for the overlay and the overlay itself.
+
+  private overlays: SortedMap<DocumentKey, Overlay> = new SortedMap<
+    DocumentKey,
+    Overlay
+  >(DocumentKey.comparator);
+  private overlayByBatchId: Map<number, Set<DocumentKey>> = new Map<
+    number,
+    Set<DocumentKey>
+  >();
+
+  getOverlay(
+    transaction: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<Overlay | null> {
+    return PersistencePromise.resolve(this.overlays.get(key));
+  }
+
+  private saveOverlay(
+    transaction: PersistenceTransaction,
+    largestBatchId: number,
+    mutation: Mutation | null
+  ): void {
+    if (mutation === null) {
+      return;
+    }
+
+    // Remove the association of the overlay to its batch id.
+    const existing = this.overlays.get(mutation.key);
+    if (existing !== null) {
+      // @ts-ignore: we know the result of `get` is not undefined.
+      this.overlayByBatchId.get(existing.largestBatchId).delete(mutation.key);
+    }
+
+    this.overlays = this.overlays.insert(
+      mutation.key,
+      new Overlay(largestBatchId, mutation)
+    );
+
+    // Create the association of this overlay to the given largestBatchId.
+    if (this.overlayByBatchId.get(largestBatchId) === undefined) {
+      this.overlayByBatchId.set(largestBatchId, new Set<DocumentKey>());
+    }
+    // @ts-ignore: we know the result of `get` is not undefined.
+    this.overlayByBatchId.get(largestBatchId).add(mutation.key);
+  }
+
+  saveOverlays(
+    transaction: PersistenceTransaction,
+    largestBatchId: number,
+    overlays: Map<DocumentKey, Mutation>
+  ): PersistencePromise<void> {
+    overlays.forEach((mutation, key) => {
+      this.saveOverlay(transaction, largestBatchId, mutation);
+    });
+    return PersistencePromise.resolve();
+  }
+
+  removeOverlaysForBatchId(
+    transaction: PersistenceTransaction,
+    batchId: number
+  ): PersistencePromise<void> {
+    if (this.overlayByBatchId.has(batchId)) {
+      const keys = this.overlayByBatchId.get(batchId);
+      this.overlayByBatchId.delete(batchId);
+      // @ts-ignore: we know keys is not undefined
+      for (const key of keys) {
```

Comment:
```suggestion
      const keys = this.overlayByBatchId.get(batchId)!;
      this.overlayByBatchId.delete(batchId);
      for (const key of keys) {
```

You should ideally just do one get() though instead of has()/get();